Model/MetaDataModel.php

```diff
@@ -0,0 +1,184 @@
+     * @return string
+     *
+     * @since 1.2.0
+     */
+    protected function _getFieldShortName($sLongFieldName)
+    {
+        return str_replace("{$this->_sCoreTable}__", '', $sLongFieldName);
+    }
+
+    /**
+     * Extends the `load` method to also load meta fields from the database.
+     *
+     * @inheritdoc
+     * @param string $sOxid
+     * @return bool
+     *
+     * @since 1.2.0
+     */
+    public function load($sOxid)
+    {
+        $bIsLoaded = is_subclass_of($this, BaseModel::class) ? parent::load($sOxid) : true;
+
+        foreach ($this->loadMetaData() as $sFieldShortName => $mFieldValue) {
+            $sFieldLongName = $this->_getFieldLongName($sFieldShortName);
+
+            $this->$sFieldLongName = new Field($mFieldValue);
+        }
+
+        return $bIsLoaded;
+    }
+
+    /**
+     * Extends the `save` method to also save meta fields to the database.
+     *
+     * @inheritdoc
+     * @return string|bool
+     *
+     * @since 1.2.0
+     */
+    public function save()
+    {
+        $aMetaData = [];
+
+        foreach ($this->getMetaFields() as $sFieldLongName => $oField) {
+            $sFieldShortName = $this->_getFieldShortName($sFieldLongName);
+
+            if ($this->$sFieldLongName instanceof Field) {
+                $aMetaData[$sFieldShortName] = $this->$sFieldLongName->getRawValue();
+            }
+        }
+
+        $this->deleteMetaData();
+        $this->saveMetaData($aMetaData);
+
+        return is_subclass_of($this, BaseModel::class) ? parent::save() : true;
+    }
+
+    /**
+     * Returns all meta fields attached to the object.
+     *
+     * @return array
+     *
+     * @since 1.2.0
+     */
+    public function getMetaFields()
+    {
+        return array_filter(get_object_vars($this), function ($sKey) {
+            return strpos($sKey, "{$this->_sCoreTable}__") === 0 &&
+                !isset($this->_aFieldNames[$this->_getFieldShortName($sKey)]);
+        }, ARRAY_FILTER_USE_KEY);
+    }
+
+    /**
+     * Loads meta data from the database.
+     *
+     * @return array
+     *
+     * @since 1.2.0
+     */
+    public function loadMetaData()
+    {
+        $aMetaData = [];
+        $aRows = DatabaseProvider::getDb(DatabaseProvider::FETCH_MODE_ASSOC)->select(
+            "SELECT `KEY`, `VALUE` FROM `{$this->getTableName()}` WHERE `OXOBJECTID` = '{$this->getId()}'"
+        )->fetchAll();
+
+        foreach ($aRows as $aRow) {
+            $aMetaData[$aRow['KEY']] = unserialize($aRow['VALUE']);
+        }
+
+        return $aMetaData;
+    }
+
+    /**
+     * Saves meta data to the database.
+     *
+     * @param array $aMetaData Associative array of meta data to save.
+     * @return integer
+     *
+     * @since 1.2.0
+     */
+    public function saveMetaData($aMetaData = [])
+    {
+        if (!$aMetaData) {
+            return 0;
+        }
+
+        // delete all duplicate keys for this object
+        $this->deleteMetaData(array_keys($aMetaData));
```

Comment:
Why is there duplicated keys? is it not an update then?

Reply:
Theoretically you could add multiple meta data entries with the same key. To prevent this, we delete keys before saving them. So yes, in this case it would be an update.

Reply:
Same as the one above? Why not Update?

Reply:
I changed it so that `INSERT` is only executed if the key doesn't already exist for the object, otherwise an `UPDATE` is executed.

Reply:
Great